pkg/asset/installconfig/aws/metadata.go

```diff
@@ -63,3 +67,75 @@ func (m *Metadata) AvailabilityZones(ctx context.Context) ([]string, error) {
 
 	return m.availabilityZones, nil
 }
+
+// PrivateSubnets retrieves subnet metadata indexed by subnet ID, for
+// subnets that the cloud-provider logic considers to be private
+// (i.e. not public).
+func (m *Metadata) PrivateSubnets(ctx context.Context) (map[string]Subnet, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	err := m.populateSubnets(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.privateSubnets, nil
+}
+
+// PublicSubnets retrieves subnet metadata indexed by subnet ID, for
+// subnets that the cloud-provider logic considers to be public
+// (e.g. with suitable routing for hosting public load balancers).
+func (m *Metadata) PublicSubnets(ctx context.Context) (map[string]Subnet, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	err := m.populateSubnets(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.publicSubnets, nil
+}
+
+func (m *Metadata) populateSubnets(ctx context.Context) error {
+	if len(m.publicSubnets) > 0 || len(m.privateSubnets) == 0 {
+		return nil
+	}
+
+	if len(m.subnets) == 0 {
+		return errors.New("no subnets configured")
+	}
+
+	session, err := m.unlockedSession(ctx)
+	if err != nil {
+		return err
+	}
+
+	m.vpc, m.privateSubnets, m.publicSubnets, err = subnets(ctx, session, m.subnets)
+	return err
+}
+
+// VPC retrieves the VPC ID containing PublicSubnets and PrivateSubnets.
+func (m *Metadata) VPC(ctx context.Context) (string, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if m.vpc == "" {
+		if len(m.subnets) == 0 {
+			return "", errors.New("cannot calculate VPC without configured subnets")
+		}
+
+		session, err := m.unlockedSession(ctx)
+		if err != nil {
+			return "", err
+		}
+
+		m.vpc, err = subnetVPC(ctx, session, m.subnets[0])
```

Comment:
why not call populatesubnets to make sure the vpc is set https://github.com/openshift/installer/pull/2477/commits/c7c4967e1623550fafc7b2aa66b9931f9a63fd7f#diff-db7720e4268d007137ad43473cb2b398R115

Reply:
> why not call populatesubnets to make sure the vpc is set

Because all the hassle of fetching route tables is unnecessary in this case.  But it doesn't really matter, because this only gets called late in `pkg/asset/cluster/tfvars.go`, so rerolled to use `populateSubnets` with 72c533bbb -> b98a00d07.